File: backend/Api.EscolaIdiomas/Api.EscolaIdiomas.Domain/Services/Alunos/AlunosService.cs
using Api.EscolaIdiomas.Domain.DTO.Requests.Alunos;
using Api.EscolaIdiomas.Domain.DTO.Responses.Alunos;
using Api.EscolaIdiomas.Domain.Interfaces.Alunos;
using Api.EscolaIdiomas.Domain.Models.Alunos;

namespace Api.EscolaIdiomas.Domain.Services.Alunos
{
    public class AlunosService : IAlunosService
    {

        private readonly IAlunosRepository _alunosRepository;

        public AlunosService(IAlunosRepository alunosRepository)
        {
            _alunosRepository = alunosRepository;
        }

        public async Task DeleteAluno(long id)
        {
            var aluno = await _alunosRepository.GetAlunoById(id);

            if (aluno == null)
            {
                throw new Exception($"Nenhum aluno encontrado com id: {id}");
            }


            await _alunosRepository.DeleteAluno(id);
        }

        public async Task<GetAlunoByIdResponse> GetAlunoById(long id)
        {
            var aluno = await _alunosRepository.GetAlunoById(id);

            if (aluno == null) return null;


            return new GetAlunoByIdResponse()
            {
                Id = aluno.Id,
                Nome = aluno.Nome,
                Sobrenome = aluno.Sobrenome,
                DataDeNascimento = aluno.DataDeNascimento,
                DataMatricula = aluno.DataMatricula,
                Email = aluno.Email,
                Telefone = aluno.Telefone,
                Ativo = aluno.Ativo,
            };
        }

        public async Task<IEnumerable<GetAlunosResponse>> GetAlunos()
        {
            var alunos = await _alunosRepository.GetAlunos();

            if (alunos == null)
            {
                throw new Exception($"Nenhum aluno encontrado!");
            }

            var response = alunos.Select(a => new GetAlunosResponse 
            { 
                Id = a.Id, 
                Nome = a.Nome,
                Sobrenome = a.Sobrenome,
                Email = a.Email,
                Telefone = a.Telefone,
                DataDeNascimento = a.DataDeNascimento,
                DataMatricula = a.DataMatricula,
                Ativo = a.Ativo
            });

            return response;
        }

        public async Task<InsertAlunoResponse> InsertAluno(InsertAlunoRequest request)
        {
            try
            {
                var aluno = new Aluno()
                {
                    Nome = request.Nome,
                    Sobrenome = request.Sobrenome,
                    DataDeNascimento = request.DataDeNascimento,
                    Email = request.Email,
                    Telefone = request.Telefone,
                    Ativo = request.Ativo,
                    DataMatricula = request.DataMatricula ?? DateTime.Now // Usar data atual se não fornecida
                };

                var response = await _alunosRepository.InsertAluno(aluno);

                var newAluno = new InsertAlunoResponse()
                {
                    Id = response
                };

                return newAluno;
            }
            catch (Exception)
            {
                throw;
            }
        }

        public async Task UpdateAluno(long id, UpdateAlunoRequest request)
        {
            try
            {
                var aluno = await _alunosRepository.GetAlunoById(id);

                if (aluno == null)
                {
                    throw new Exception($"Não foi possivel encontrado um aluno com o id {id}");
                }

                aluno.Telefone = request.Telefone;
                aluno.Email = request.Email;


                await _alunosRepository.UpdateAluno(aluno);
            }
            catch (Exception ex)
            {

                throw new Exception($"Erro ao atualizar aluno: {ex}");
            }
        }
    }
}

File: frontend/src/services/alunosService.ts
import { useApi } from "../composable/api";
import type { CreateAlunoFormData } from "../components/forms/schemas/alunoSchema";

// Interface para a resposta da API
export interface InsertAlunoResponse {
  id: number;
}

// Interface para a requisição da API
export interface InsertAlunoRequest {
  nome: string;
  sobrenome: string;
  dataDeNascimento: string; // ISO date string
  email: string;
  telefone: string;
  dataMatricula: string; // ISO date string
  ativo: boolean;
}

// Interface para resposta de erro da API
export interface AlunoError {
  mensagem: string;
  status: string;
}

// Hook personalizado para serviços de alunos
export const useAlunosService = () => {
  const { post, get, put, del, loading, error, clearError } = useApi();

  // Função para inserir um novo aluno
  const insertAluno = async (data: CreateAlunoFormData): Promise<InsertAlunoResponse> => {
    try {
      // Preparar dados para a API
      const requestData: InsertAlunoRequest = {
        nome: data.nome,
        sobrenome: data.sobrenome,
        dataDeNascimento: data.dataDeNascimento.toISOString(),
        email: data.email,
        telefone: data.telefone,
        dataMatricula: new Date().toISOString(), // Data atual como data de matrícula
        ativo: data.ativo
      };

      console.log("Enviando dados para API:", requestData);

      // Fazer chamada para a API usando o composable
      const response = await post<InsertAlunoResponse>("/alunos", requestData);
      
      console.log("Aluno cadastrado com sucesso:", response.data);
      
      return response.data;
      
    } catch (error: any) {
      console.error("Erro ao cadastrar aluno:", error);
      
      // Re-throw o erro para ser tratado pelo componente
      throw error;
    }
  };

  // Função para buscar todos os alunos
  const getAlunos = async () => {
    try {
      const response = await get("/alunos");
      return response.data;
    } catch (error: any) {
      console.error("Erro ao buscar alunos:", error);
      throw error;
    }
  };

  // Função para buscar aluno por ID
  const getAlunoById = async (id: number) => {
    try {
      const response = await get(`/alunos/${id}`);
      return response.data;
    } catch (error: any) {
      console.error(`Erro ao buscar aluno com ID ${id}:`, error);
      throw error;
    }
  };

  // Função para atualizar aluno
  const updateAluno = async (id: number, data: Partial<CreateAlunoFormData>) => {
    try {
      const requestData = {
        ...data,
        dataDeNascimento: data.dataDeNascimento?.toISOString(),
      };

      const response = await put(`/alunos/${id}`, requestData);
      return response.data;
    } catch (error: any) {
      console.error(`Erro ao atualizar aluno com ID ${id}:`, error);
      throw error;
    }
  };

  // Função para deletar aluno
  const deleteAluno = async (id: number) => {
    try {
      const response = await del(`/alunos/${id}`);
      return response.data;
    } catch (error: any) {
      console.error(`Erro ao deletar aluno com ID ${id}:`, error);
      throw error;
    }
  };

  return {
    // Estados
    loading,
    error,
    clearError,
    
    // Métodos
    insertAluno,
    getAlunos,
    getAlunoById,
    updateAluno,
    deleteAluno,
  };
};

// Funções utilitárias para uso direto (sem hook) - usando a instância da API diretamente
import { api } from "../composable/api";

export const alunosService = {
  // Função para inserir aluno sem usar hook
  insertAluno: async (data: CreateAlunoFormData): Promise<InsertAlunoResponse> => {
    const requestData: InsertAlunoRequest = {
      nome: data.nome,
      sobrenome: data.sobrenome,
      dataDeNascimento: data.dataDeNascimento.toISOString(),
      email: data.email,
      telefone: data.telefone,
      dataMatricula: new Date().toISOString(),
      ativo: data.ativo
    };

    const response = await api.post<InsertAlunoResponse>("/alunos", requestData);
    return response.data;
  },

  // Função para buscar todos os alunos sem usar hook
  getAlunos: async () => {
    const response = await api.get("/alunos");
    return response.data;
  },

  // Função para buscar aluno por ID sem usar hook
  getAlunoById: async (id: number) => {
    const response = await api.get(`/alunos/${id}`);
    return response.data;
  },

  // Função para atualizar aluno sem usar hook
  updateAluno: async (id: number, data: Partial<CreateAlunoFormData>) => {
    const requestData = {
      ...data,
      dataDeNascimento: data.dataDeNascimento?.toISOString(),
    };

    const response = await api.put(`/alunos/${id}`, requestData);
    return response.data;
  },

  // Função para deletar aluno sem usar hook
  deleteAluno: async (id: number) => {
    const response = await api.delete(`/alunos/${id}`);
    return response.data;
  },
};
